cmd/erasure-object.go

```diff
@@ -424,7 +440,27 @@ func (er erasureObjects) getObjectFileInfo(ctx context.Context, bucket, object s
 	// Read metadata associated with the object from all disks.
 	metaArr, errs := readAllFileInfo(ctx, disks, bucket, object, opts.VersionID, readData)
 
-	readQuorum, _, err := objectQuorumFromMeta(ctx, metaArr, errs, er.defaultParityCount)
+	readQuorum, _, err := func() (int, int, error) {
+		readQuorum, writeQuorum, versionIDHeal, err := objectQuorumFromMeta(ctx, metaArr, errs, er.defaultParityCount)
+		if err != nil && errors.Is(err, errErasureReadQuorum) {
+			if versionIDHeal != "" && opts.VersionID == "" {
+				// VersionIDHeal is present means it is inconsistent
+				// and when an exact versionID is not specified, we perform heal
+				// based resolution.
+				_, herr := er.HealObject(ctx, bucket, object, versionIDHeal, madmin.HealOpts{
+					NoLock: true,
```

Comment:
The reason locking would work is that WRITEs are blocked during READs of course you can have as many reads as necessary - since xl.meta is self-contained here.

So no one is really overriding the data here and Heal only heals relevant parts anyways.